src/actions/receive-profile.js

```diff
@@ -434,49 +493,46 @@ async function _fetchProfile({ url, onTemporaryError }: FetchProfileArgs) {
 export function retrieveProfileFromStore(
   hash: string
 ): ThunkAction<Promise<void>> {
-  return async function(dispatch) {
-    dispatch(waitingForProfileFromStore());
-
-    try {
-      const serializedProfile = await _fetchProfile({
-        url: `https://profile-store.commondatastorage.googleapis.com/${hash}`,
-        onTemporaryError: (e: TemporaryError) => {
-          dispatch(temporaryErrorReceivingProfileFromStore(e));
-        },
-      });
-
-      const profile = unserializeProfileOfArbitraryFormat(serializedProfile);
-      if (profile === undefined) {
-        throw new Error('Unable to parse the profile.');
-      }
-
-      dispatch(receiveProfileFromStore(profile));
-    } catch (error) {
-      dispatch(fatalErrorReceivingProfileFromStore(error));
-    }
-  };
+  return retrieveProfileOrZipFromUrl(
+    `https://profile-store.commondatastorage.googleapis.com/${hash}`
+  );
 }
 
-export function retrieveProfileFromUrl(
+/**
+ * Runs a fetch on a URL, and downloads the file. If it's JSON, then it attempts
+ * to process the profile. If it's a zip file, it tries to unzip it, and save it
+ * into the store so that the user can then choose which file to load.
+ */
+export function retrieveProfileOrZipFromUrl(
```

Comment:
There are a couple or arguments I can throw in here:

1. I think it's reasonable to do sniffing when dealing with local files, and less reasonable when dealing with remote files. Having 2 different functions here to deal with a zip file or with a json profile doesn't prevent doing some sniffing when dealing with local files in the drag&drop code.

2. Once you do it it's not possible to move back and we'll have to maintain this sniffing at this location forever (because we'll have to support this URL scheme forever). However if we introduce a new URL format for remote zip files, it would be easy to change it later.

If these arguments still don't work for you, please at least split the function in 2 parts: one for json handling, one for zip handling:

```javascript
if (isProbablyZipFile(response)) {
  return {
    zip: await handleZipFile(response),
  }
} else {
  return {
    profile: await handleJson(response),
  };
}
```

Error handling would be done in these `handleXXX` function.

Reply:
Ok, I'll take another pass at it.